### PR TITLE
refactor(license): use goyacc for license parser

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
Refactor https://github.com/aquasecurity/trivy/pull/3335

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/3267

## Related PRs
- [x] https://github.com/aquasecurity/trivy/pull/3335

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/licensing/expression/types.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+var versioned = []string{
+	licensing.AGPL10,
+	licensing.AGPL30,
+	licensing.GPL10,
+	licensing.GPL20,
+	licensing.GPL30,
+	licensing.LGPL20,
+	licensing.LGPL21,
+	licensing.LGPL30,
+}
```
Comment (DmitriyLewen) [inline pkg/licensing/expression/types.go:20]: SPDX docs contain `GFDL` licenses with `-or-later/-only` suffix(e.g. GFDL-1.1-invariants-only/GFDL-1.1-invariants-or-later).
`licenseclassifier` doesn't contain these licenses in types(https://github.com/google/licenseclassifier/blob/7c62d6fe8d3aa2f39c4affb58c9781d9dc951a2d/license_type.go#L24-L177).
But we can take licenses not only from `licenseclassifier`(e.g. dpkg copyright files).

Looks like we also need to add their.

File: pkg/licensing/expression/lexer.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+type Lexer struct {
+	s      *bufio.Scanner
+	result Expression
+	errs   error
+}
+
+func NewLexer(reader io.Reader) *Lexer {
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		// cf. bufio.ScanWords()
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/licensing/expression/lexer.go]: Updated
https://github.com/aquasecurity/trivy/pull/3824/commits/0d23b03a27baacaa8b990b7a64b58e763339b3b4

File: pkg/sbom/spdx/marshal.go [hunk 3]
```diff
@@ -362,11 +364,16 @@ func GetLicense(p ftypes.Package) string {
...
-		licensing.Normalize,
-		expression.NormalizeForSPDX,
-	)
+	license := strings.Join(lo.Map(p.Licenses, func(license string, index int) string {
+		return fmt.Sprintf("(%s)", license)
+	}), " AND ")
+	s, err := expression.Normalize(license, licensing.Normalize, expression.NormalizeForSPDX)
+	if err != nil {
+		// Not fail on the invalid license
+		log.Logger.Warnf("Unable to parse licenses %q", license)
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/sbom/spdx/marshal.go]: Fixed
https://github.com/aquasecurity/trivy/pull/3824/commits/7bf3f18eaed428e7a3924d7019018f77fb4c1ce6

File: pkg/licensing/expression/types.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	left := c.left.String()
+	if _, ok := c.left.(CompoundExpr); ok {
+		left = fmt.Sprintf("(%s)", left)
+	}
+	right := c.right.String()
+	if _, ok := c.right.(CompoundExpr); ok {
+		right = fmt.Sprintf("(%s)", right)
+	}
+	return fmt.Sprintf("%s %s %s", left, c.conjunction.literal, right)
+}
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/licensing/expression/types.go]: I didn't have a good idea yesterday, but I came up with an idea today.
https://github.com/aquasecurity/trivy/pull/3824/commits/632ee5ebfe0e1b93e8db1296348fa98c507a1b41

File: pkg/licensing/expression/types.go (hunk 1 above)
Comment (knqyf263) [reply pkg/licensing/expression/types.go]: I knew it, but I was lazy 😄 
https://github.com/aquasecurity/trivy/pull/3824/commits/c38b5fafb7fe1792ca3d50c0dfe70e08e45cb68f

**Review by DmitriyLewen — APPROVED** [review]

**Pull request merged**